perf(ivy): remove megamorphic read from renderStringify

crisbeto opened the pull request:
The `renderStringify` function is used in a lot of performance-sensitive places, however it contains a megamorphic read which is used primarily for error messages. These changes introduce a new function that can be used to stringify output for errors and removes the megamorphic read from `renderStringify`.

This PR resolves FW-1286.

JoostK commented on packages/core/src/render3/util/misc_utils.ts:
```diff
@@ -25,17 +22,31 @@ export function isDifferent(a: any, b: any): boolean {
+ * used for error messages.
+ */
+export function stringifyForError(value: any) {
+  if (typeof value == 'object' && typeof value.type == 'function') {
```
Could you make this safe to be called with `null`?

crisbeto replied:
Good point. I don't think we have any code paths that would call this with null, but it's good to have it covered anyway.

gkalpak commented on packages/core/src/render3/util/misc_utils.ts:
```diff
@@ -25,17 +22,31 @@ export function isDifferent(a: any, b: any): boolean {
+ * used for error messages.
+ */
+export function stringifyForError(value: any) {
+  if (typeof value == 'object' && value != null && typeof value.type == 'function') {
```
OOC, any reason not to use strict equality?

crisbeto replied:
I don't think so. I just moved the logic over from the old function, but it might be a good idea to switch it now since we're changing it anyway.

crisbeto committed:
perf(ivy): remove megamorphic read from renderStringify

The `renderStringify` function is used in a lot of performance-sensitive places, however it contains a megamorphic read which is used primarily for error messages. These changes introduce a new function that can be used to stringify output for errors and removes the megamorphic read from `renderStringify`.

This PR resolves FW-1286.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:245046171:BASE:245046197:1556118479704:fb9d76bd)

AndrewKushnir commented:
@crisbeto could you please rebase your PR on top of the latest master and resolve conflict(s), which appear after merging other changes to master? Thank you.

crisbeto commented:
Rebased.
